anndata/tests/helpers.py

```diff
@@ -256,8 +256,14 @@ def gen_adata(
         awkward_ragged=gen_awkward((12, None, None)),
         # U_recarray=gen_vstr_recarray(N, 5, "U4")
     )
-    with warnings.catch_warnings():
-        warnings.simplefilter("ignore", ExperimentalFeatureWarning)
+
+    ctx = (
+        pytest.warns(ExperimentalFeatureWarning)
+        if AwkArray in {*obsm_types, *varm_types}
+        else nullcontext()
+    )
+
+    with ctx:
```

Comment:
```suggestion
    with warnings.catch_warnings():
        warnings.filterwarnings(
            "ignore",
            "Support for Awkward Arrays is currently experimental",
            ExperimentalFeatureWarning,
        )
```

I don't think this function should error if the warning isn't thrown. We have tests specifically for this, and I don't think every test that uses `gen_adata` with an awkward array needs to fail if it doesn't throw a warning.

I do think it could make sense to be more specific with the feature warning though.